runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -878,4 +887,198 @@ public void shutdown() {
+
+        @Default
+        @Exclude
+        private final boolean withSchema = true;
+
+        @NonNull
+        @Exclude
+        private final TableOptions tableOptions;
+
+        public static <K extends Message, V extends Message, M extends Message>
+        TableDescriptor<K, V, M> build(Class<K> kClass, Class<V> vClass, Class<M> mClass, TableOptions tableOpts) {
+            return TableDescriptor
+                    .<K, V, M>builder()
+                    .kClass(kClass)
+                    .vClass(vClass)
+                    .mClass(mClass)
+                    .withSchema(mClass != null)
+                    .tableOptions(tableOpts)
+                    .build();
+        }
+
+        public static <K extends Message, V extends Message, M extends Message>
+        TableDescriptor<K, V, M> build(Class<K> kClass, Class<V> vClass, Class<M> mClass) {
+            TableOptions defaultOpts;
+            try {
+                defaultOpts = TableOptions.fromProtoSchema(vClass);
+            } catch (Exception e) {
+                throw new IllegalStateException(e);
+            }
+
+            return build(kClass, vClass, mClass, defaultOpts);
+        }
+
+        public SchemaOptions getSchemaOptions() {
+            return tableOptions.getSchemaOptions();
+        }
+
+        public V getDefaultValueMessage() {
+            return getDefaultMessage(vClass);
+        }
+
+        public K getDefaultKeyMessage() {
+            return getDefaultMessage(kClass);
+        }
+
+        private static <T> T getDefaultMessage(Class<T> clazz) {
+            try {
+                return ClassUtils.cast(clazz.getMethod(DEFAULT_INSTANCE_METHOD_NAME).invoke(null));
+            } catch (Exception ex) {
+                throw new IllegalStateException(DEFAULT_METHOD_NOT_FOUND_ERR_MSG);
+            }
+        }
+
+        public void metadataHandler(Consumer<M> defaultMetadataHandler) {
+            if (mClass == null) {
+                return;
+            }
+
+            M defaultMessage = getDefaultMetadataMessage();
+            defaultMetadataHandler.accept(defaultMessage);
+        }
+
+        private M getDefaultMetadataMessage() {
+            if (mClass == null) {
+                return null;
+            }
+
+            return getDefaultMessage(mClass);
+        }
+
+        public Object[] getArgs() {
+            if (withSchema) {
+                return new Object[]{getIndexer()};
+            } else {
+                return new Object[]{};
+            }
+        }
+
+        public ProtobufIndexer getIndexer() {
+            SchemaOptions schemaOptions = getSchemaOptions();
+            V msg = getDefaultValueMessage();
+            return new ProtobufIndexer(msg, schemaOptions);
+        }
+    }
+
+    @Builder
+    @EqualsAndHashCode
+    @ToString
+    public static class FullyQualifiedTableName {
+        @Default
+        private final Optional<String> namespace = Optional.empty();
```

Comment:
namespace should never be empty - can you please share what is the usecase for having an Optional here?

Reply:
Unfortunately your statement is not true, at least for our internal code and tests: https://github.com/CorfuDB/CorfuDB/blob/424a38a137bcddeadce72cc7e2b46d2b979766d5/runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java#L528

Reply:
This is our internal code which is being cautious and checking for null, I am just curious to know which layer actually passes in null for namespace - is it Compactor? is it LogReplicator? is it CorfuServer? or some other place?

Reply:
As you can see there is an "if" in getFullyQualifiedTableName that checks if namespace is null or an empty string, it suggest that namespace could be empty. So, my changes just follows the rules that are already there, I just made it more clear and enforced to have the check if namespace is empty or not, rather than just having if condition which checks if namespace is null. 
In case if we want to enforce that the namespace has to be NOT null, I'd happy to do that, but in a next PR, right now I just want to follow whatever rules out there